Isaac/load-verse-of-the-day

ittullos opened the pull request:
This commit adds a function to the Verse model that returns the verse of the day. At the moment I have only been given 100 verses by the client. I am handling this by dividing the day of the year by 100 and using the remainder to select the verse of the day. This is not ideal as it will repeat the verses every 100 days. Eventually when I have all 366 verses this will need to be changed. I'm also having a little trouble with the test. It passes when I run it in isolation but fails sometimes when I run the whole suite. I would also love some feedback on best practices for naming of my tests and methods. Just kinda taking a stab at it right now.

ittullos committed:
Add verse_of_the_day method to Verse model

This commit adds a method to the Verse model that returns the verse
of the day. At the moment I have only been given 100 verses by the
client. I am handling this by dividing the day of the year by 100 and
using the remainder to select the verse of the day. This is not ideal
as it will repeat the verses every 100 days. Eventually when I have all
366 verses this will need to be changed.

zbarnard1 commented on spec/models/verse_spec.rb:
```diff
@@ -58,4 +58,12 @@
     end
   end
+
+  describe 'verse of the day' do
```
This generally would be `describe '.verse_of_the_day'` because it's testing a method. Note the dot, it's a class method rather than an instance method (you'd use `#verse_of_the_day` in that case.

ittullos replied:
Awesome!! Thanks!

zbarnard1 commented on spec/models/verse_spec.rb:
```diff
@@ -58,4 +58,12 @@
   end
+
+  describe 'verse of the day' do
+    it 'returns the correct verse' do
```
I can get this to fail pretty readily. The reason it fails is because the test assumes you only have 100 Verses, but if other tests run first it increments the sequence before you create the list and the Verse day starts on 7 or 8.

ittullos replied:
Okay that makes sense. Probably going to test it by creating a single verse first if that will work.

zbarnard1 replied:
This works. It looks to me like we'll never hit Verse #1, since we always add 1 to the current day. This test fails:
```ruby
describe '.verse_of_the_day' do
  include ActiveSupport::Testing::TimeHelpers

  it 'returns the correct verse' do
    # year, month, day, hour, minute, second
    travel_to Time.local(2024, 1, 1, 1, 0, 0) do
      verse = create(:verse, day: 1)
      expect(Verse.verse_of_the_day.scripture).to eq(verse.scripture)
    end
  end
end
```

zbarnard1 replied:
Ahh got it. Let me update my example and I'll be right back.

zbarnard1 replied:
```ruby
it 'returns the correct verse' do
  # year, month, day, hour, minute, second
  travel_to Time.local(2024, 4, 9, 1, 0, 0) do
    verse = create(:verse, day: 1)

    expect(Verse.verse_of_the_day.scripture).to eq(verse.scripture)
  end

  travel_to Time.local(2024, 4, 10, 1, 0, 0) do
    verse = create(:verse, day: 2)

    expect(Verse.verse_of_the_day.scripture).to eq(verse.scripture)
  end
end
```

ittullos replied:
Okay awesome! Makes perfect sense!! Thanks so much for your help! 🙏

ittullos commented:
Okay so I think I've fixed the test. I just decided to create a single verse like you recommended. Let me know if everything looks up to snuff!

zbarnard1 commented on spec/models/verse_spec.rb:
```diff
@@ -58,4 +58,13 @@
+      date = (Date.today.yday % 100) + 1
+      verse = create(:verse, day: date)
+
+      expect(Verse.verse_of_the_day.scripture).to eq(verse.scripture)
```
You could drop the `scripture` part here, you already have the verse as a variable here so you can just `expect(Verse.verse_of_the_day).to eq(verse)`

ittullos replied:
Ahh okay makes sense!! 🙌

ittullos committed:
Delete old integration tests

ittullos committed:
Seed 100 verses from excel file

ittullos committed:
Load verse of the day from /home endpoint.

ittullos commented:
Wrestled with this for a while but I was able to seed the database with verses and then load the verse of the day from my React Native application! One thing I did have to change was the way the user email was accessed from the backend. Instead of accessing the header with the key "P4L-email", I had to use "HTTP_P4L_EMAIL". I guess it capitalizes whatever key I give it and adds that HTTP_ to the front. Let me know if there's anything that needs to be fixed! Thanks!

zbarnard1 commented on db/seeds.rb:
```diff
@@ -0,0 +1,18 @@
+require 'roo'
+require_relative '../app/models/verse'
+
+Load 100 verses from an .xlsx document into the Verses table
```
Should this be a comment?

ittullos replied:
Yep whoops 😅

zbarnard1 commented on db/seeds.rb:
```diff
@@ -0,0 +1,18 @@
+  Verse.create(scripture: verse_text, notation: notation, version: 'CSB', day: i - 1)
+end
+
+User.create(email: "isaac.tullos@gmail.com", first_name: "Isaac", last_name: "Tullos")
```
Won't this raise an error if it's run more than once? We might want to split seeding a User out from the Verses, and they can be run separately or together. One approach is to create classes that you can call from a rake task:
```ruby
  # rake db:seed:users
  namespace :db do
    namespace :seed do
      desc "Seed Users from /db/seeds/users.yml"
      task :users => :environment do
        SeedUsers.run
      end
    end
  end
```

Then you have the code for that in the /lib/tasks/dv_seed_users.rake file:
```ruby
class SeedUsers
  def run
    er.create(email: "isaac.tullos@gmail.com", first_name: "Isaac", last_name: "Tullos")
  end 
```

Rake tasks can call other tasks, so you can have something that seeds everything, or just specific models.
https://www.blandersoft.com/short/rake-yaml-rails-seed/

ittullos replied:
I actually meant to ask a question about this, so good looking out! I had just been commenting out whatever I didn't want to run, but splitting it into different tasks makes so much more sense. I'll go ahead and fix this before I merge!

zbarnard1 commented on spec/requests/app_spec.rb:
```diff
@@ -12,4 +12,28 @@
+
+      end
+
+      travel_to Time.local(2024, 4, 10, 1, 0, 0) do
```
⛏️ Tiny nitpick here, you've already tested the logic of the verse of the day so you don't need to repeat it here, we're more concerned with the API response in this context, which is covered by the 200 status and the data returned by the first one.

ittullos replied:
Gotcha! 🫡

zbarnard1 reviewed (changes requested):
We're close! The API stuff looks good for the most part. You could add some examples for an un-authenticated user, but we'll be back in that area shortly so it's not a blocker right now.
You could still update your commit message to say method instead of function if you wanted. The biggest issue is the non-commented comment in the seed code.

zbarnard1 commented on db/seeds.rb:
```diff
@@ -0,0 +1,18 @@
+(2..101).each do |i|
+  notation = sheet.cell(i, 1)
+  verse_text = sheet.cell(i, 2)
+  Verse.create(scripture: verse_text, notation: notation, version: 'CSB', day: i - 1)
```
⛏️ Shouldn't this version attribute come from the file, too?

ittullos replied:
So for now the only verses they gave me were in CSB version. Eventually we will want to allow them to select which version they want to use. Maybe I should go ahead and add the version to the 3rd column so that way I don't have to fix it later?

zbarnard1 reviewed (approved):
Feel free to just fix up that missing comment and then merge when you're ready!

ittullos committed:
Test seed tasks

Wrote tests for user and verse rake tasks. This required adding a new
migration to add a unique index to the verses table. This also meant
that some tests needed to be re-written, particularly the tests
in app_spec.rb.

zbarnard1 committed:
Add missing require in verse spec

This was causing a test failure when the file was run on its own.

zbarnard1 committed:
Remove db/seeds.rb

This task can still be run, even though the file for it no longer exists
in source control.

zbarnard1 reviewed (approved):
LGTM!
